qcodes_contrib_drivers/drivers/Attocube/ANC350.py

```diff
@@ -0,0 +1,510 @@
+﻿from qcodes import Instrument, Parameter
+from qcodes.instrument.channel import InstrumentChannel, ChannelList
+from qcodes.utils.validators import Numbers
+from qcodes_contrib_drivers.drivers.Attocube.ANC350Lib import ANC350LibActuatorType, ANC350v3Lib, ANC350v4Lib
+from typing import Any, Callable, Dict, Optional, Union
+
+
+class Anc350Axis(InstrumentChannel):
+    """
+    Representation of an axis of the ANC350
+
+    The Attocube ANC350 has 3 axis, one for every direction.
+
+    Args:
+        parent: the Instrument that the channel is attached to
+        name: the name of the axis itself
+        axis: the index of the axis (0..2)
+
+    Attributes:
+        position: Get the current position on a single axis
+        frequency: Set the frequency of the output signal. Depending on positioner type and usage of other axes one can
+            adjust the frequency from 1Hz up to 5kHz (only on one axis at one time is a frequency above 2kHz allowed)
+        amplitude: Value for the drive voltage of the piezo drive. Bychanging this value, the step size of the
+            positioner can be varied. Value for the drive voltage of the piezo drive. Bychanging this value, the step
+            size of the positioner can be varied.
+        status:
+        voltage: Sets the DC level on the voltage output when no sawtooth based motion and no feedback loop
+            is active.
+        target_position: Sets the target position for automatic motion (start_auto_move). For linear type actuators the
+            position unit is m, for goniometers and rotators it is degree.
+        target_range: Defines the range around the target position where the target is considered to be reached.
+        actuator: Selects the actuator to be used for the axis from actuator presets.
+        actuator_name: Get the name of the currently selected actuator
+        capacitance: Performs a measurement of the capacitance of the piezo motor and returns the result. If no
+            motor is connected, the result will be 0.
+            The function doesn't return before the measurement is complete; this will take a few seconds of time.
+    """
+
+    def __init__(self, parent: "ANC350", name: str, axis: int):
+        super().__init__(parent, name)
+
+        self._axis = axis
+
+        self.add_parameter("position",
+                           label="Position",
+                           get_cmd=self._get_position,
+                           unit="mm or m°")
```

Comment:
The position unit is um or m°. Is there any reason to use mm?

Reply:
The DLL returns ° or m, depending on the current actuator type (see `anc350res.h` in [inc.zip](https://github.com/QCoDeS/Qcodes_contrib_drivers/files/5408772/inc.zip)). Our customer preferred to use mm and m° instead, because they also use these units internally and thus do not confuse the different units.

If you prefer other units, I can change it.

Reply:
I see. It is a bit confusing, though. dll output should conform the units stated in the manual. To me, it is strange to have the position in the order of meter. Micrometer seems to be a correct unit.

Reply:
Ok, then I would suggest to use um and u°, because then I can use the same conversion factor for both units. So I don't have to check the actuator every time I want to read the position.

Otherwise I had to use a different factor for each unit when using um and m° and I am afraid that checking the actuator will go wrong and the position will be wrong by a factor of 1000.

Reply:
From the manual, which shows their manufacturer software menus, all positions are stated as um and m° with the resolution of nm and u°. It means that we can't rely on the manual for all units and units depends on the output of dll file, right? The internal resolution is 1 nm or 1 µ°, so how you could use the same conversion for linear and rotational positions?

Reply:
The DLL returns positions in units of m or °. It's some time ago, but I'm pretty sure that we've tested it. E.g. the function `_get_position` currently doesn't know, which unit was returned. It just sees the float-value.. So the best would be to use a constant factor for conversion, e.g. 10^3 to convert units into mm and m° or 10^6 to convert units into µm and µ°.

If you want to use um and m° instead, I had to use a factor of 10^6 for converting m in um and 10^3 to convert ° in m°. Of course, this is possible and not very complicated, but I am not sure if it is worth it.

The advantage would be that you have units that are the same as in LabView, but the problem would be, that we then need two function-calls to get the position (+ actuator type), which might make the driver a bit slower. And I'm afraid, what happens, when the actuator type changes during measurement. If it first was linear, the position unit is m. When reading the position, the result is multiplied by 10^6 to convert it into µm. If the actuator type changes and the unit is now °, it has to be mutliplied by 10^3 to convert it into m°. But what if the actuator changes during measurement and the driver doesn't notice. Then it might happen that it uses the wrong conversion factor and returns mm or µ°, what is not expected by the user.

Additionally, I think it is a bit overkill to do so much background conversions, just to have other units. Since the parameters are given as floats anyway, the internal precision of nm and µ° can still be achieved.

But if you prefer using µm and m°, I can implement it like this. I just wanted you to know my concerns.

Reply:
OK. I am not saying we have to have specific units, although it is recommended to conform to the manual. The only point is that the instrument sets/ gets correct location and resolution. So if you have tested the instrument already and you know what the returned numbers and your correct conversion are, then you can program it like that and inform users which units are used as you did in the code.

The important point is to define correct resolution for 2 parameters that can be set: 'Target Range' and 'Target Position'. You should be careful as you used 1e-6 for both units.